components/webgpu/lib.rs

```diff
@@ -296,20 +314,69 @@ impl WGPU {
             receiver,
             sender,
             script_sender,
-            global: wgpu::hub::Global::new("wgpu-core", factory),
+            global: wgpu::hub::Global::new("wgpu-core", factory, wgt::BackendBit::PRIMARY),
             adapters: Vec::new(),
             devices: Vec::new(),
             _invalid_adapters: Vec::new(),
+            buffer_maps: HashMap::new(),
             webrender_api: webrender_api_sender.create_api(),
             webrender_document,
             external_images,
             wgpu_image_map,
         }
     }
 
-    fn run(mut self) {
+    fn run(&'a mut self) {
         while let Ok(msg) = self.receiver.recv() {
+            self.devices.iter().for_each(|device| {
+                let global = &self.global;
+                let device_id = device.0;
+                gfx_select!(device_id => global.device_poll(device_id, false));
+            });
             match msg {
+                WebGPURequest::BufferMapAsync {
+                    sender,
+                    buffer_id,
+                    host_map,
+                    map_range,
+                } => {
+                    let map_info = BufferMapInfo {
+                        buffer_id,
+                        sender,
+                        global: &self.global,
+                        size: (map_range.end - map_range.start) as usize,
+                    };
+                    self.buffer_maps.insert(buffer_id, map_info);
+
+                    unsafe extern "C" fn callback(status: BufferMapAsyncStatus, userdata: *mut u8) {
+                        let info = (userdata as *mut BufferMapInfo).as_ref().unwrap();
+                        match status {
+                            BufferMapAsyncStatus::Success => {
+                                let global = info.global;
+                                let data_pt = gfx_select!(info.buffer_id =>
+                                    global.buffer_get_mapped_range(info.buffer_id, 0, None));
+                                let data = slice::from_raw_parts(data_pt, info.size).to_vec();
```

Comment:
are we sending an immutable slice? That would only work for read mapping, not write mappings.

Reply:
I am a bit confused about what your concern here is. We are sending a vector to the client. If I have to change the mutability I'll have to do it on the client-side while receiving, i.e. `WebGPUResponse::BufferMapAsync(mut bytes)` in `handle_response()`? And we'll be writing back any data in the `WebGPURequest::Unmap` via `write_buffer()` or 'set_buffer_sub_data()`. How does that change things here?

Reply:
Here is how I see this working, at this point at least:
  1. for a write-mappable buffer, the implementation carries around the shared memory with its data. When user asks for mapping, you make the server to notify the client when the GPU is done, and you just expose this shmem to the user to modify, followed by `set_buffer_sub_data` on the server side.
  2. for a read-mappable buffer, the server waits for GPU, copies the immutable slice out into a shmem that is given to the user.

I suppose the code we see here is for (2) only? In this case, why do we have `host_map` parameter to the message?

Reply:
I see. I had a different approach in my mind. I was going with sending vectors across the processes for data exchange between client and server.
I guess you are pointing towards using [`IpcSharedMemory`](https://docs.rs/ipc-channel/0.14.1/ipc_channel/ipc/struct.IpcSharedMemory.html) (similar to the previous `MapReadAsync` implementation).

Reply:
Sorry. I see there is a bigger concern here than the `IpcSharedMemory`. But it would probably be solved when I figure out the intersecting ArrayBuffers part.